WFLY-8473 Add capabilites to undertow subsystem

ctomc opened the pull request:
Documentation of capabilities PR https://github.com/wildfly/wildfly-capabilities/pull/13

- WFLY-8439 - Unregistering node throws error: MEM: Can't update or insert context
- JBEAP-9971 - Cannot read 'referencing-deployments' in /subsystem=undertow/application-security-domain
- WFLY-7576 minimize deprecated API usage in undertow subsystem 
- partial WFLY-7987 work - improve usability of errors when working trough CLI
- move to service loader backed transformers registration
- JBEAP-6462 Setting elytron ssl-context on undertow without reload

bstansberry commented on mod_cluster/extension/src/main/java/org/wildfly/extension/mod_cluster/ContainerEventHandlerAdapterBuilder.java:
```diff
@@ -1,8 +1,9 @@
 
 public interface ContainerEventHandlerAdapterBuilder {
-    ServiceBuilder<?> build(ServiceTarget target, String connector, int statusInterval);
+    ServiceBuilder<?> build(ServiceTarget target, OperationContext context, String connector, int statusInterval);
```
@pferraro  is this ok?

ctomc replied:
this is needed to be able to resolve capability later on.

pferraro replied:
The module that implements ContainerEventHandlerAdapterBuilder knows nothing of mod_cluster operations or its model, thus passing the OperationContext is not appropriate here.  It should instead use CapabilityServiceSupport.

This will eventually be converted to use CapabilityServiceBuilder[1].
e.g.
CapabilityServiceBuilder<?> createBuilder(String connector, int statusInterval);
[1] https://github.com/wildfly/wildfly/blob/master/clustering/common/src/main/java/org/jboss/as/clustering/controller/CapabilityServiceBuilder.java

ctomc replied:
there is already CapabilityServiceBuilder in core nowdays and most of this PR uses it.
UndertowEventHandlerAdapterBuilder which is the only impl of this interface is also changed as part of this PR, so that shouldn't be an issue

pferraro replied:
I'm just looking at your new CapabilityServiceBuilder now.  I suspect this might allow me to simplify the CapabilityServiceBuilder from wildfly-clustering-common to something like:
```
interface CapabilityBuilder<T> extends Builder<T> {
    @Override
    CapabilityServiceBuilder<T> build(CapabilityServiceTarget target);
}
```
My initial point still stands though.  ContainerEventHandlerAdapterBuilder.build(...) should accept a CapabilityServiceSupport rather than an OperationContext.

pferraro replied:
Actually, I should think you can change it to:
`ServiceBuilder<?> build(CapabilityServiceTarget target, String connector, int statusInterval);`

bstansberry commented:
@ctomc please push up a PR to wildfly-capabilities that includes all the caps in the undertow subsystem. Please pay particular attention to the "PRIVATE:" settings in that PR, as described in "Private Capabilities" in https://github.com/wildfly/wildfly-capabilities/blob/master/README.md.

This seems generally ok, although CI says differently. :) But this is a huge change and I want to be able to focus on what the actual capabilities are; what the contracts are. The wildfly-capabilities PR will let me do that.

bstansberry commented on undertow/src/main/resources/META-INF/services/org.jboss.as.controller.transform.ExtensionTransformerRegistration:
```diff
@@ -0,0 +1,63 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
```
This is repeated multiple times.

ctomc replied:
fixed

ctomc committed:
WFLY-8473 undertow capabilities

WFLY-8439 - Unregistering node throws error: MEM: Can't update or insert context
JBEAP-9971 - Cannot read 'referencing-deployments' in /subsystem=undertow/application-security-domain
WFLY-7576 minimize deprecated API usage in undertow subsystem
partial WFLY-7987 work - improve usability of errors when working trough CLI
move to service loader backed transformers registration
JBEAP-6462 Setting elytron ssl-context on undertow without reload

darranl commented:
Is updating the copyright year on all these classes the correct thing to do?

ctomc commented:
@darranl I do think so.

rhusar commented:
So many different and unrelated changes/issues in on commit. Please don't do that :-(

rhusar commented:
Review the modcluster changes -- now with the CapabilityServiceSupport it looks fine. 👍

bstansberry reviewed (approved):
Ok, assuming the nearly done, currently passing CI jobs pass.

Going forward refactoring/cleanup changes must be done in separate commits to make reviewing the core change practical.

kabir commented:
Retest this please

pferraro commented:
I think this commit broke mod_cluster:
```
2017-03-31 20:51:45,216 ERROR [org.jboss.as.controller.management-operation] (Controller Boot Thread) WFLYCTL0013: Operation ("add") failed - address: ([
    ("subsystem" => "undertow"),
    ("server" => "default-server"),
    ("host" => "default-host"),
    ("setting" => "http-invoker")
]) - failure description: {
    "WFLYCTL0412: Required services that are not installed:" => ["org.wildfly.mod_cluster"],
    "WFLYCTL0180: Services with missing/unavailable dependencies" => ["org.wildfly.undertow.http-invoker.host.default-host.http-invoker is missing [org.wildfly.mod_cluster]"]
}
2017-03-31 20:51:45,233 INFO  [org.jboss.as.server] (Controller Boot Thread) WFLYSRV0010: Deployed "negotiationAuthenticator.war" (runtime-name : "negotiationAuthenticator.war")
2017-03-31 20:51:45,235 INFO  [org.jboss.as.controller] (Controller Boot Thread) WFLYCTL0183: Service status report
WFLYCTL0184:    New missing/unsatisfied dependencies:
      service org.wildfly.mod_cluster (missing) dependents: [service org.wildfly.undertow.http-invoker.host.default-host.http-invoker] 

```

rhusar commented:
Actually, this didn't break mod_cluster but it has broken http-invoker.

Here is a quick PR to restore the dependency introduced by original Stuart's PR https://github.com/wildfly/wildfly/pull/9885 if anyone needs the fix during weekend. I am not yet sure why the dependency is needed, if this is not just a workaround we will have to define a proper contract (exposing the container event handler as a mod_cluster capability, where the implementation is provided by the mod_cluster-undertow service).

ctomc commented:
@pferraro @rhusar Found the issue, there is a clash between service names, org.widfly.mod_cluster vs org.wildfly.modcluster 
It would also be avoided if mod cluster properly moved to capabilities as well.

rhusar commented:
Are you sure? 
https://github.com/wildfly/wildfly/pull/9858/files#diff-8fdcb0b77f2875d1184e1dfa324f7abbR59
https://github.com/wildfly/wildfly/pull/9858/files#diff-a6d49c7daf65f77b4095fc808bfd5340R72
The problem is what I described in the above problem.

ctomc commented:
@pferraro @rhusar fix https://github.com/wildfly/wildfly/pull/9895
